Add ID scraper

chriszs opened the pull request:
Add Idaho scraper.

## Steps to test
```sh
python -m warn.cli ID
```

Closes #82

chriszs commented:
Deals with gaps caused by merged cells, like this:

<img width="1401" alt="Screen Shot 2022-01-20 at 6 58 52 AM" src="https://user-images.githubusercontent.com/1103467/150335355-1ebdf3a3-bbb2-4757-954e-882ed7986cd2.png">

By filling down like this:

<img width="1309" alt="Screen Shot 2022-01-21 at 7 57 20 PM" src="https://user-images.githubusercontent.com/1103467/150618068-cdcdfe4e-27ae-4ff0-b8b5-edefe0cc0b49.png">

Notice we're excluding the number of employees column to preserve the merged cell total and avoid double counting.

chriszs commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,123 @@
+            if (
+                clean_text == ""
+                and row_index > 0
+                and col_index < len(output_rows[row_index - 1])
```
Black and flake8 disagree about whether the `and` should go before the line or after, [this website says flake8 is wrong](https://www.flake8rules.com/rules/W503.html).

chriszs replied:
Okay, upon researching this I realized that the linter was also complaining when I put these at the end of the line. How was it possible it didn't like _both_? And why does the flake8 maintainer say this has rule is ignored? Answer, using `ignore` in the flake8 config overrides default ignore rules, including this one. [This StackOverflow answer](https://stackoverflow.com/a/68172142/1871478) recommends `extend-ignore` instead, which is what this patch now modifies the flake8 config to do.

chriszs commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,123 @@
+
+    cache_key = f"{cache_state}/WARNNotice.pdf"
+
+    # verify=False because there's a persistent cert error
+    # we're working around.
```
This results in a nasty warning, but as far as I can tell there may be no good way around it. I looked into whether the problem was not having `certifi` as [StackOverflow indicated](https://stackoverflow.com/questions/52805115/certificate-verify-failed-unable-to-get-local-issuer-certificate), but it seemed like that wasn't it.

chriszs replied:
Leaving this as-is.

chriszs commented:
This one's ready for review, @palewire.

palewire commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,127 @@
+
+    # verify=False because there's a persistent cert error
+    # we're working around.
+    response = requests.get(url, verify=False)
```
What do you think about adding a verify keyword option to our `utils.get_url` utility so we can use it here? We'd get the logging, and make the utility stronger too.

chriszs replied:
Yes, I think that's a good idea. I'll do it.

palewire commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,127 @@
+    # verify=False because there's a persistent cert error
+    # we're working around.
+    response = requests.get(url, verify=False)
+    with open(cache_key, "wb") as file:
+        file.write(response.content)
```
Could we use the built-in `cache.write` function from the instance of our Cache class?

chriszs replied:
I tried to, but I think `cache.write` doesn't handle binary files. It only works for strings. The better approach here I guess would be to extend `cache.download` to take `verify` and use that.

chriszs replied:
Added.

palewire commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,127 @@
+        for index, page in enumerate(pdf.pages):
+            rows = page.extract_table()
+
+            output_rows = output_rows + _clean_table(rows, index)
```
```suggestion
            output_rows +=  _clean_table(rows, index)
```
I've actually never written a line like this, but would it work?!

palewire commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,127 @@
+
+    # Write out the data to a CSV
+    data_path = Path(data_dir, f"{state_code}.csv")
+    utils.write_rows_to_csv(output_rows, data_path)
```
```suggestion
    utils.write_rows_to_csv(data_path, output_rows)
```
This got reversed upstream. Need to catch up.

palewire commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,127 @@
+            output_rows = output_rows + _clean_table(rows, index)
+
+    # Write out the data to a CSV
+    data_path = Path(data_dir, f"{state_code}.csv")
```
```suggestion
    data_path = data_dir / f"{state_code}.csv"
```
My picky thing.

palewire commented on warn/scrapers/id.py:
```diff
@@ -0,0 +1,127 @@
+                clean_text == ""
+                and row_index > 0
+                and col_index < len(output_rows[row_index - 1])
+                and output_rows[0][col_index] != "No. of Employees Affected"
```
Would you mind breaking this out into an `_is_header` true/false test so it's a little clear to newbies what the test is doing. (Assume I am clear what it's doing!)

chriszs replied:
It doesn't do that, but refactored into functions that say what it is doing.

palewire reviewed (changes requested):
Thanks for the creative use pdfplumber here. Big picture it all looks great. I just have a couple small things I've flagged. Let me know what you think.

chriszs commented:
Thanks for the suggestions! Refactored to use `cache.download` (and added a feature to that to pass verify or any other kwarg), added some util functions to make that one if statement test read clearer, integrated your suggestions and cleaned everything up.

palewire commented:
Thanks! You're in!
